python/metatensor-learn/metatensor/learn/nn/equivariant_transform.py

```diff
@@ -0,0 +1,142 @@
+from typing import List, Optional, Union
+
+import torch
+from torch.nn import Module
+
+from .._backend import Labels, TensorMap
+from .._dispatch import int_array_like
+from ._utils import _check_module_map_parameter
+from .module_map import ModuleMap
+
+
+class EquivariantTransform(Module):
+    """
+    A custom :py:class:`torch.nn.Module` that applies an arbitrary shape- and
+    equivariance-preserving transformation to an input :py:class:`TensorMap`.
+
+    ``module`` is passed as a callable with parameters ``in_features`` and optionally
+    ``dtype`` and ``device``. This callable constructs an arbitrary shape-preserving
+    :py:class:`torch.nn.Module` transformation (i.e. :py:class:`torch.nn.Tanh`).
+    Separate instances are created for each block using the metadata information
+    passed in ``in_keys`` and ``in_features``.
+
+    For invariant blocks in `in_keys` and indexed by `invariant_keys`, the
+    transformation is applied as is. For covariant blocks, an invariant multiplier is
+    created applying the transformation to the norm of the block over the component
+    dimension.
+
+    :param module: a :py:class:`list` of :py:class:`torch.nn.Module` containing the
+        transformations to be applied to each block indexed by
+        :param in_keys:. Transformations for invariant and covariant blocks differ. See
+        above.
+    :param in_keys: :py:class:`Labels`, the keys that are assumed to be in the input
+        :py:class:`TensorMap` in the :py:meth:`forward` method.
+    :param in_features: :py:class:`list` of :py:class:`int`, the number of features in
+        the input tensor for each block indexed by the keys in :param in_keys:. If
+        passed as a single value, the same number of features is assumed for all blocks.
+    :param out_properties: :py:class:`list` of :py:class`Labels` (optional), the
+        properties labels
+        of the output. By default the output properties are relabeled using
+        Labels.range.
+    :param invariant_keys: a :py:class:`Labels` object that is used to select the
+        invariant keys from ``in_keys``. If not provided, the invariant keys are assumed
+        to be those where key dimensions ``["o3_lambda", "o3_sigma"]`` are equal to
+        ``[0, 1]``.
+    """
```

Comment:
It would be nice to have some usage example for this class, since I doubt someone will be able to use it just by reading the docstring.

Reply:
In the docstring or elsewhere? i.e. could we make a tutorial of using `learn.nn` modules together in defining custom architectures separately?

Reply:
Ok we've added a docstring example, tutorial example coming later :)